### ISSUE TITLE
Can not connect with 127.0.0.1

### ISSUE THREAD

**Issue by batyrchary** [intel/sgx-ra-sample#21]
Hello,

I was not able to make remote attestation between two PCs.

I provided input parameters as specified but no success.

I think problem is in socket connection it is not working with IP addresses, it connects only to localhost but when I provide 127.0.0.1 as IP address or IP of my remote server  it can not make connection.

I tried different ports.
I turned off firewall. 
But still was not able to make connection.

What do you think might be the problem?
Any help would be appreciated.
Thanks

**Comment by jmechalas** [comment]
What error do you see in the logs when using 127.0.0.1 as the address? The sample calls getaddrinfo(3) to resolve the peer name, and that function _should_ accept both names and IP addresses. The debug output will help me see what it thinks it's connecting to.

Make sure you have debug output on. Also please be careful not to include passwords in the output as the debugging statements will reveal those.

**Comment by batyrchary** [comment]
Thank you for you response

on client log I only see 

---- Client Log Timestamp --------------------------------------------------
2019-03-29 10:46:01
----------------------------------------------------------------------------
Mar 29 2019 10:46:02 127.0.0.1: Mar 29 2019 10:46:02 connect: failed on error 0



I tried to debug it on my own 
inside of constructor MsgIO::MsgIO(const char *peer, const char *port)

client cant connect and this if condition doesnt hold and break from loop 
if (connect(s, addr->ai_addr, (int)addr->ai_addrlen) == 0)

**Comment by jmechalas** [comment]
OK. I'll spend some quality time with this today and see what sort of weird logic bug we have in there.

**Comment by jmechalas** [comment]
What OS are you on?

I am not able to reproduce this on Linux. Running it on the command line:

```
johnm@host:~/git/sgx-ra-sample$ ./run-client 127.0.0.1
+++ using default public key

---- Msg0 Details ----------------------------------------------------------
Extended Epid Group ID: 00000000
----------------------------------------------------------------------------

---- Msg1 Details ----------------------------------------------------------
msg1.g_a.gx = 16a49b13d6c03092f55a6d433eb0a4142e896bb4b1a8df2d9b5e1baf9a601945
msg1.g_a.gy = 287cfeb59e47656055f3efc3a83526e2cf524fe5dc3d2274c0b23123f65f895d
msg1.gid    = ee0a0000
----------------------------------------------------------------------------

---- Copy/Paste Msg0||Msg1 Below to SP -------------------------------------
0000000016a49b13d6c03092f55a6d433eb0a4142e896bb4b1a8df2d9b5e1baf9a601945287cfeb59e47656055f3efc3a83526e2cf524fe5dc3d2274c0b23123f65f895dee0a0000
----------------------------------------------------------------------------
Waiting for msg2
[...]
```

The server log shows:

```
Listening for connections on port 7777
Waiting for a client to connect...
Connection from 127.0.0.1
Waiting for msg0||msg1
+++ read 145 bytes from socket

---- read buffer -----------------------------------------------------------
0000000016a49b13d6c03092f55a6d433eb0a4142e896bb4b1a8df2d9b5e1baf9a601945287cfeb59e47656055f3efc3a83526e2cf524fe5dc3d2274c0b23123f65f895dee0a0000
----------------------------------------------------------------------------

---- Msg0 Details (from Client) --------------------------------------------
msg0.extended_epid_group_id = 0
----------------------------------------------------------------------------

---- Msg1 Details (from Client) --------------------------------------------
msg1.g_a.gx = 16a49b13d6c03092f55a6d433eb0a4142e896bb4b1a8df2d9b5e1baf9a601945
msg1.g_a.gy = 287cfeb59e47656055f3efc3a83526e2cf524fe5dc3d2274c0b23123f65f895d
msg1.gid    = ee0a0000
[...]
```

**Comment by batyrchary** [comment]
I am using Windows10.

I think issue is ipv4 and ipv6 but I am not sure.

**Comment by jmechalas** [comment]
I'm out of the office on Monday but I'll bring up my windows development system on Tuesday and dig into this further.

**Comment by batyrchary** [comment]
Thanks for your help.

**Comment by jmechalas** [comment]
I have confirmed this issue on Windows 10 when using 127.0.0.1 as the server name/IP.

**Comment by jmechalas** [comment]
You are correct that it is binding using IPv6 addresses, but is not accepting a connection when an IPv4 address is supplied. I am digging into the details of getaddrinfo().

**Comment by jmechalas** [comment]
OK. I have a fix for this but I need to validate that it's not breaking anything under Linux. While Windows supports dual-stack sockets, the default with Vista and later is for an IPv6 socket to only accept connections from IPv6 addresses. I have addressed that and will push the new code once validation is done.

**Comment by jmechalas** [comment]
I have posted a fix to master and tagged it as v2.2.3. I also fixed an issue with setting SO_REUSEADDR in the wrong place while I was at it.

**Comment by batyrchary** [comment]
Now it seems to work as expected, I think we can close this issue.

Thanks for your help,
Batyr
